core/src/frame_lifecycle.rs

```diff
@@ -0,0 +1,169 @@
+    ///
+    /// Frame scripts are the AS3 equivalent of old-style `DoAction` tags. They
+    /// run when a movie clip enters a given timeline frame.
+    FrameScripts,
+
+    /// We're about to finish frame processing.
+    ///
+    /// When we exit a completed frame, we fire `exitFrame` on the broadcast
+    /// list.
+    Exit,
+
+    /// We're about to destroy children of existing display objects.
+    ///
+    /// All `RemoveObject` tags should execute at this time.
+    Destroy,
+
+    /// We're not currently executing any frame code.
+    ///
+    /// At this point in time, event handlers are expected to run. No frame
+    /// catch-up work should execute.
+    Idle,
+}
+
+/// Run one frame according to AVM1 frame order.
+pub fn run_all_phases_avm1<'gc>(context: &mut UpdateContext<'_, 'gc, '_>) {
+    // In AVM1, we only ever execute the update phase, and all the work that
+    // would ordinarily be phased is instead run all at once in whatever order
+    // the SWF requests it.
+    *context.frame_phase = FramePhase::Update;
+
+    // AVM1 execution order is determined by the global execution list, based on instantiation order.
+    for clip in context.avm1.clip_exec_iter() {
+        if clip.removed() {
+            // Clean up removed objects from this frame or a previous frame.
+            // Can be safely removed while iterating here, because the iterator advances
+            // to the next node before returning the current node.
+            context.avm1.remove_from_exec_list(context.gc_context, clip);
+        } else {
+            clip.run_frame(context);
+        }
+    }
+
+    // Fire "onLoadInit" events.
+    context
+        .load_manager
+        .movie_clip_on_load(context.action_queue);
+
+    *context.frame_phase = FramePhase::Idle;
+}
+
+/// Run one frame according to AVM2 frame order.
+pub fn run_all_phases_avm2<'gc>(context: &mut UpdateContext<'_, 'gc, '_>) {
+    let stage = context.stage;
+
+    *context.frame_phase = FramePhase::Destroy;
+    stage.destroy_frame(context);
+
+    *context.frame_phase = FramePhase::Enter;
+    stage.enter_frame(context);
+
+    *context.frame_phase = FramePhase::Construct;
+    stage.construct_frame(context);
+    stage.frame_constructed(context);
+
+    *context.frame_phase = FramePhase::Update;
+    stage.run_frame_avm2(context);
+
+    *context.frame_phase = FramePhase::FrameScripts;
+    stage.run_frame_scripts(context);
+
+    *context.frame_phase = FramePhase::Exit;
+    stage.exit_frame(context);
+
+    *context.frame_phase = FramePhase::Idle;
+}
+
+/// Run all previously-executed frame phases on a newly-constructed display
+/// object.
+///
+/// This is a no-op on AVM1, which has it's own catch-up logic.
+pub fn catchup_display_object_to_frame<'gc>(
+    context: &mut UpdateContext<'_, 'gc, '_>,
+    dobj: DisplayObject<'gc>,
+) {
+    match (*context.frame_phase, context.avm_type()) {
+        (_, AvmType::Avm1) => {}
+        (FramePhase::Enter, AvmType::Avm2)
```

Comment:
why do we catch-up to `construct_frame` on Enter phase? This will cause the second construct after catching up, right? - or is this expected?
(for the record, if I'm wrong, it probably just means that this needs an extra line of comment to not confuse others :D )

Reply:
So, the thing that motivated this is the fact that event handlers (which run at `Idle`) can construct clips and expect to see frame one's children ready to go. My gut feeling is that it's probably wrong to skip `construct_frame` on *any* frame phase. However, I do not have any test coverage for this right now - removing `construct_frame` from `Enter` doesn't fail any tests.

Reply:
So, after adding in a bunch of tests for gotos on every frame phase I could think of, I can confidently say that yes, we do actually need to `construct_frame` on `Enter`.